Automigration: Remove `docs.autodocs` field

ndelangen opened the pull request:
## What I did

- Implemented `removeDocsAutodocs` fix to check and remove the `docs.autodocs` field from the main configuration, which is no longer supported in Storybook 9.
- Added tests for the new migration functionality to ensure correct behavior during the check and run phases.
- Updated the main fixes index to include the new migration.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Implemented a migration fix to remove the deprecated 'docs.autodocs' field from Storybook's main configuration, which was deprecated in v7-8 and removed in v9.

- Added `removeDocsAutodocs` fix in `code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts` to handle configuration cleanup
- Added comprehensive test coverage in `remove-docs-autodocs.test.ts` for both check and run phases
- Fix preserves other docs properties while removing autodocs field, or removes entire docs object if autodocs was the only property
- Includes dry run support and proper error handling for configuration updates



<!-- /greptile_comment -->

ndelangen committed:
Add migration to remove deprecated docs.autodocs field from Storybook config

- Implemented `removeDocsAutodocs` fix to check and remove the `docs.autodocs` field from the main configuration, which is no longer supported in Storybook 9.
- Added tests for the new migration functionality to ensure correct behavior during the check and run phases.
- Updated the main fixes index to include the new migration.

yannbf commented on code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts:
```diff
@@ -0,0 +1,79 @@
+
+  prompt: () => {
+    return dedent`
+      The docs.autodocs field in your Storybook configuration has been removed in Storybook 9.
+      This field was deprecated in Storybook 7-8 and is no longer supported.
+      
+      We will remove this field from your configuration.
+      
+      Would you like us to:
+      1. Remove the docs.autodocs field from your Storybook configuration
```
```suggestion
      The \`docs.autodocs\` field in Storybook's main configuration has been removed in Storybook 9.
      This field was deprecated in Storybook 7-8 and is no longer supported.
      
      We can make the necessary changes in your configuration automatically.
      
      More info: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#mainjs-docsautodocs-is-deprecated
```

yannbf commented on code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts:
```diff
@@ -0,0 +1,79 @@
+    try {
+      let hasAutodocs = false;
+
+      await updateMainConfig({ mainConfigPath, dryRun: true }, async (main) => {
+        const docs = main.getFieldValue(['docs']) || {};
+        hasAutodocs = 'autodocs' in docs;
+      });
```
Can you use loadConfig instead? The updateMainConfig utility is to used to write into the file instead.

yannbf reviewed (changes requested):
The automigration should check the value of autodocs, whether it's `'tag'`, `true` or `false`.

`'tag'` or `false` -> Just remove the field from main.js
`true` -> Remove the field and update the `preview.js` file to use `tags: ["autodocs"]`.

You can access the `previewConfigPath` as part of the `run` options

ndelangen committed:
Update code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

ndelangen committed:
Enhance removeDocsAutodocs migration to handle autodocs values

- Updated the `removeDocsAutodocs` fix to capture and process the `autodocs` field, allowing for both boolean and 'tag' values.
- Modified the run phase to update the preview configuration with tags when autodocs is set to true.
- Expanded test cases to cover new scenarios for autodocs detection and removal, ensuring comprehensive validation of the migration logic.

ndelangen committed:
Merge branch 'norbert/automigration-remove-autodocs' of https://github.com/storybookjs/storybook into norbert/automigration-remove-autodocs

ndelangen committed:
Add test for handling autodocs tag in empty tags array

- Introduced a new test case to verify that the `removeDocsAutodocs` migration correctly adds the 'autodocs' tag to the tags array in `preview.js` when the autodocs option is enabled.
- Enhanced the mock configuration setup to simulate the conditions for this scenario, ensuring comprehensive coverage of the migration logic.

yannbf commented on code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts:
```diff
@@ -0,0 +1,95 @@
+
+    try {
+      const config = loadConfig(mainConfigPath).parse();
+      const docs = config.getFieldValue(['docs']) || {};
```
I _think_ you can get rid of two variables and do a safer access with:
`const autodocs = config.getSafeFieldValue(['docs', 'autodocs'])`

which will not crash in case the user has some dynamic code that might break the AST parsing, and it will return undefined in that case.

Then you just use `autodocs` instead of `autodocs` + `hasAutodocs`

yannbf commented on code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts:
```diff
@@ -0,0 +1,95 @@
+
+      if (!dryRun) {
+        // Only add autodocs tag if it's not already present
+        if (!tags.includes('autodocs')) {
```
Can you add a log prior to this, saying something like `Updating the docs parameter in the preview config file`, else the automigration won't show messages that explain what's going on. Also flip the if like
```
      if (!tags.includes('autodocs')) {
        // Only add autodocs tag if it's not already present
        if (!dryRun) {
          previewConfig.setFieldValue(['tags'], [...tags, 'autodocs']);
          await writeConfig(previewConfig);
        } else {
          // the log message here
        }
      }
```

ndelangen committed:
Refactor removeDocsAutodocs migration to use getSafeFieldValue

- Updated the `removeDocsAutodocs` migration to replace `getFieldValue` with `getSafeFieldValue` for retrieving the `autodocs` field from the configuration.
- Removed the `hasAutodocs` property from the result object, simplifying the migration logic.
- Adjusted test cases to reflect the changes in the migration's handling of the `autodocs` field, ensuring accurate validation of the updated logic.

ndelangen committed:
Refactor logging in removeDocsAutodocs migration

- Updated the logging logic in the `removeDocsAutodocs` migration to ensure that the 'autodocs' tag is only logged when it is being added to the preview config.
- Adjusted the condition for adding the 'autodocs' tag to prevent unnecessary logging during a dry run, enhancing clarity in migration output.

ndelangen committed:
Refactor logging in removeDocsAutodocs migration

- Simplified the logging logic to ensure the 'autodocs' tag update is only logged when necessary, enhancing clarity in migration output.
- Removed redundant logging during the dry run phase, improving the overall efficiency of the migration process.

ndelangen committed:
Merge branch 'next' into norbert/automigration-remove-autodocs

ndelangen committed:
Refactor removeDocsAutodocs migration to streamline dry run handling

- Updated the `removeDocsAutodocs` migration to only execute the main configuration update when not in dry run mode, improving efficiency and clarity.
- Simplified the logic for removing the `autodocs` field from the `docs` object while maintaining the existing functionality.

ndelangen committed:
Merge branch 'norbert/automigration-remove-autodocs' of https://github.com/storybookjs/storybook into norbert/automigration-remove-autodocs

ndelangen committed:
Refactor removeDocsAutodocs migration to improve dry run handling and logging

- Adjusted the logic to ensure the main configuration update is executed regardless of the dry run mode, enhancing clarity and efficiency.
- Updated logging to only occur when the 'autodocs' tag is added to the preview config, preventing unnecessary output during dry runs.

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.ts:
```diff
@@ -0,0 +1,96 @@
+
+    // If autodocs was true, update preview config to use tags
+    if (autodocs === true && previewConfigPath) {
+      const previewConfig = loadConfig(previewConfigPath).parse();
```
I think the naming of `updateMainConfig` is a bit unlucky, because you can pass any kind of config into it, also the preview config. Would you mind changing it and to use it also here to adjust the preview configuration?

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/remove-docs-autodocs.test.ts:
```diff
@@ -0,0 +1,339 @@
+  Fix<{ autodocs: boolean | 'tag' | undefined }>
+>;
+
+describe('remove-docs-autodocs migration', () => {
```
The tests are super cryptic, but I understand why you have chosen to do it this way.

I would recommend an alternative approach, though to make it more readable for humans: Instead of mocking `updateMainConfig` I would rather prefer to mock `readFile (node:fs)` and `writeFile (node:fs)`. Then, your assertions will be much simplified, because you can just make a inline snapshot of the transformed configuration and the assertions are actually pretty easy to grasp and it is easy to understand how the config file looked before and after the migration.

ndelangen committed:
Refactor removeDocsAutodocs migration tests and implementation

- Updated tests to utilize `readFile` and `writeFile` mocks for improved file handling.
- Simplified the test structure by removing unnecessary mock configurations and directly using `readFile` for configuration retrieval.
- Adjusted the migration implementation to replace `loadConfig` with `readConfig`, enhancing asynchronous handling of configuration files.
- Improved logging and handling of the 'autodocs' tag in the preview configuration, ensuring clarity and efficiency in the migration process.

ndelangen committed:
Merge branch 'next' into norbert/automigration-remove-autodocs

valentinpalkovic reviewed (approved):
So much better! LGTM!

ndelangen committed:
Update snapshot in removeDocsAutodocs tests to replace double quotes with single quotes for consistency

ndelangen committed:
Update removeDocsAutodocs test to use replaceAll for quote replacement in snapshot

ndelangen committed:
Merge branch 'next' into norbert/automigration-remove-autodocs
